### PR TITLE
SE-1168/FAL-92 implement ORA process summary report csv

### PR DESCRIPTION
**Pull request by swalladge**
This PR implements a summary report designed to be serialized to csv for users' progress in ORAs. The aim is to provide a report that can be used by instructors to perform tasks such as a list of students have submitted a response but not completed the peer assessment step.

It is designed to be used from the instructor dashboard, following the same process as the existing "ORA Data Report" csv generation. See PR on edx/edx-platform: https://github.com/edx/edx-platform/pull/20955

**JIRA tickets**: [OSPR-3721](https://openedx.atlassian.net/browse/OSPR-3721)

**Dependencies**: None

**Sandbox URL**:

-  LMS: https://pr20955.sandbox.opencraft.hosting/
-  Studio: https://studio.pr20955.sandbox.opencraft.hosting/

**Merge deadline**: None

**Screenshots**:

Example csv export:

```
block_name,student_id,status,is_peer_complete,is_peer_graded,is_self_complete,is_self_graded,is_training_complete,is_training_graded,num_peers_graded,num_graded_by_peers,is_staff_grade_received,is_final_grade_received,final_grade_points_earned,final_grade_points_possible
block-v1:edX+DemoX+Demo_Course+type@openassessment+block@b24c33ea35954c7889e1d2944d3fe397,e72ab704db2919f77281a30915de83cb,peer,0,1,0,0,,,1,2,0,0,,
block-v1:edX+DemoX+Demo_Course+type@openassessment+block@b24c33ea35954c7889e1d2944d3fe397,7d5d1b02d0270dfdcf3bc4ec426c1f8d,peer,0,1,0,0,,,0,2,0,0,,
block-v1:edX+DemoX+Demo_Course+type@openassessment+block@b24c33ea35954c7889e1d2944d3fe397,7a9579168b65a2137d5cc093a63c7da7,done,1,1,1,1,,,3,2,0,1,6,19
block-v1:edX+DemoX+Demo_Course+type@openassessment+block@b24c33ea35954c7889e1d2944d3fe397,d086f8d3a681a472061c9a75966f9123,self,1,0,0,0,,,3,1,0,0,,
```
Instructor dashboard -> Data Download section. Showing new 'Generate ORA Summary Report' button and report in progress notification:

**Testing instructions**:

see https://github.com/edx/edx-platform/pull/20955

**Author notes and concerns**:

- this is still a work in progress as the actual requirements for columns are being discussed

**Reviewers**
- [x] @pomegranited 
- [ ] edX reviewer[s] TBD

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @swalladge! I've created [OSPR-3721](https://openedx.atlassian.net/browse/OSPR-3721) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

File: openassessment/data.py
Comment (pomegranited) [inline openassessment/data.py]: Since some steps will never be complete (if they're not present in the assessment workflow, I think we need to indicate this for the row value:
* `complete` if statuses.get(step) and statuses[step]['complete']
* `NA` if not statuses.get(step)
* `''` otherwise.

(And since this is more complex, it's best moved to its own block of code, instead of done in one line.)

Also, need to ensure that the strings used here (`complete`, `NA`) are translatable -- so wrap them in `ugettext`/ `_()` calls.

File: openassessment/data.py
Comment (pomegranited) [inline openassessment/data.py]: Header names should be translatable too.

File: openassessment/data.py
Comment (pomegranited) [inline openassessment/data.py]: Instead of just using the step name here (which is confusing on its own), could you append `_done` to the name?  Will also need to translate this, so will need to look something like:

`_('is_{assessment_step_name}_done').format(assessment_step_name=step) for step in sorted(AssessmentWorkflow.STEPS)`.

File: openassessment/data.py
Comment (pomegranited) [inline openassessment/data.py]: There's some additional columns requested by the client, which should be added to the `AssessmentWorkflow.status_details()` output so it can be reported here:

* `unit_name`: the Open Responses tab can do this, so I believe that means we can too.  cf [`ora_blocks_listing_view()`](https://github.com/edx/edx-ora2/blob/028bd31f054533223e8a9df3bccd5c036e910d61/openassessment/xblock/openassessmentblock.py#L425), should be moved into an API.
* `assessment_name`: ditto ^
* `num_peers_assessed`: make [this logic from the `api.peer.has_finished_required_evaluating()`](https://github.com/edx/edx-ora2/blob/76e5a3c614a03af845229dda583380ac06636907/openassessment/assessment/api/peer.py#L478-L479) accessible directly from the `peer.api`, and include this data in the `AssessmentWorkflow.status_details()` output for the `peer` step.
* `num_assessed_by_peers`: make [this logic from the `api.peer.assessment_is_finished()`](https://github.com/edx/edx-ora2/blob/76e5a3c614a03af845229dda583380ac06636907/openassessment/assessment/api/peer.py#L86-L90) accessible directly from the `peer.api`, and include this in the  for the `peer` step.
* `peers_assessed_done`: this wasn't requested in the original report, but since we already have an API call for [has_finished_required_evaluating()](https://github.com/edx/edx-ora2/blob/76e5a3c614a03af845229dda583380ac06636907/openassessment/assessment/api/peer.py#L443), it's easy to add a boolean column for this.

File: openassessment/data.py
Comment (pomegranited) [inline openassessment/data.py]: I'm on the fence about whether all the `if...else` statements should be moved to proper blocks..  but maybe if you just moved them to named variables, so it's clear which one is which?

e.g.
```python
is_staff_grade_received = 1 if aw.staff_score_exists() else 0
is_final_grade_received = 1 if aw.status == AssessmentWorkflow.STATUS.done else 0
```

**Comment by pomegranited** [comment]
jenkins run all

File: openassessment/data.py
Comment (swalladge) [reply openassessment/data.py]: > unit_name: the Open Responses tab can do this, so I believe that means we can too. cf ora_blocks_listing_view(), should be moved into an API.

Currently the display names shown in the Open Responses tab are given to the function via the context dict from edx-platform, so that part isn't as easy unfortunately. 

> peers_assessed_done: this wasn't requested in the original report, but since we already have an API call for has_finished_required_evaluating(), it's easy to add a boolean column for this.

This should be already implemented as part of the steps complete columns.

Comment (pomegranited) [reply openassessment/data.py]: >> `unit_name`
> Currently the display names shown in the Open Responses tab are given to the function via the context dict from edx-platform, so that part isn't as easy unfortunately.

Ok, then we just need to make sure the client knows this limitation.

>> `peers_assessed_done`
> This should be already implemented as part of the steps complete columns? or does the complete status for the 'peer' step mean that the student has both received and sent enough?

Yep, the latter..  for an ORA problem to be fully "complete", a student must complete and receive a certain number of reviews.  So it would be good to add this metric to this report.

File: openassessment/workflow/models.py
Comment (pomegranited) [inline openassessment/workflow/models.py]: Please update the comment for this method to describe the fields added here.

And update the unit tests for `status_details` to test the newly added fields.

File: openassessment/workflow/models.py
Comment (pomegranited) [inline openassessment/workflow/models.py]: nit: `peers_graded_count`?

File: openassessment/data.py
Comment (pomegranited) [reply openassessment/data.py]: The ["arbitrary and ignored" number you mentioned here](https://github.com/edx/edx-ora2/pull/1250/files#diff-9da7828718b3af144f7a2603e217d725R233) is the key -- if you can pass in the actual `required_assessments` number when you call `has_finished_required_evaluating`, then that `done` number should be the `peers_assessed_done` requested here.

File: openassessment/assessment/api/peer.py
Comment (pomegranited) [inline openassessment/assessment/api/peer.py]: Needs unit tests added

File: setup.py
Comment (pomegranited) [inline setup.py]: Could you also merge the latest `master` branch and resolve the conflict with the upstream version change in `setup.py`?

File: openassessment/data.py
Comment (swalladge) [reply openassessment/data.py]: I can't work out how to get the `required_assessments` variable - that seems to be injected into the xblock context at runtime. However, I discovered I can use the two keys in the status dictionary - 'complete' corresponds to whether the student has assessed enough peers and 'graded' corresponds to whether the student has been assessed by enough peers.

Comment (pomegranited) [reply openassessment/data.py]: 👍

Comment (swalladge) [reply openassessment/data.py]: I added all the 'complete' and 'graded' columns in the report for completeness. This is the most generic. The only issue may be that the names aren't exactly clear and will need documenting (as in comment above).

**Comment by swalladge** [comment]
@pomegranited quick query about translations - should these actually be translated seeing as the other ORA csv report isn't?

File: openassessment/data.py
Comment (pomegranited) [inline openassessment/data.py]: From testing on the sandbox (see [edX_DemoX_Demo_Course_ORA_summary_2019-07-15-1025.csv](https://ocim-pr20955-sandbox-opencraft-hosting.s3.amazonaws.com/pr20955_sandbox_opencraft_hosting/grades-download/bb6d638a296059742509a0d319ddf8456b6dbf9a/edX_DemoX_Demo_Course_ORA_summary_2019-07-15-1025.csv)), the `final_grade` column reports this full JSON data string: 
```
{'submission': 2, 'created_at': datetime.datetime(2019, 7, 15, 10, 24, 8, 940437, tzinfo=<UTC>), 'points_earned': 6, 'submission_uuid': u'09357d3d-0ba4-494d-9398-172c60c9f9ba', 'annotations': [], 'student_item': 2, 'points_possible': 19}
```

So maybe we need two columns here, `final_grade_points_earned` and `final_grade_points_possible`.

And tests updated to test for this case :)

**Comment by swalladge** [comment]
@pomegranited Yep that would be easier to not need to update translations in the PR. I just did that to keep travis happy.

File: openassessment/data.py
Comment (swalladge) [reply openassessment/data.py]: Nice catch!

Comment (swalladge) [reply openassessment/data.py]: Ok, this is interesting - the existing tests *should* have a user with a final grade (according to the `test_collect_ora2_data` test), but the AssessmentWorkflow items return None. Can these be out of sync somehow? Perhaps it's related to the fact that the collect_ora2_data  method directly gets the submissions information so it has a score even if the ora workflow isn't complete. Not sure what is a more correct source of information in this context...

Comment (swalladge) [reply openassessment/data.py]: Hmm actually I think it's the way the existing test setUp sets the data - it's directly on the submission objects instead of involving the assessment workflow procedure. Might need to create a new test class.

Comment (swalladge) [reply openassessment/data.py]: Solved.

**Comment by natabene** [comment]
@swalladge Thank you for your contribution. Can you let me know once it is reedy for our review?

File: openassessment/data.py
Comment (pomegranited) [inline openassessment/data.py]: The client's new diagrams mention `current_step`. Is this something we can provide with the available data?

File: openassessment/data.py
Comment (pomegranited) [inline openassessment/data.py]: Could you provide a full description of each of the columns returned by this method, and the expected values?  

E.g., the `num_peers_graded` and `num_graded_by_peers` is a little hard to distinguish just from the header name, so would be good to have a text description of what the difference is.

File: openassessment/assessment/test/test_peer.py
Comment (swalladge) [reply openassessment/assessment/test/test_peer.py]: Hehe I can't take credit; copied from other tests. :joy:

File: openassessment/data.py
Comment (swalladge) [reply openassessment/data.py]: Agreed. Makes much more sense. :+1:

File: openassessment/data.py
Comment (swalladge) [reply openassessment/data.py]: We can provide a status from the assessment workflow. Could be a step, or a status like 'done', 'cancelled'. Should be enough. Also means the `is_cancelled` column is redundant and can be removed.

File: openassessment/data.py
Comment (pomegranited) [inline openassessment/data.py]: ... or empty string if the ORA problem workflow does not include this step.

File: openassessment/data.py
Comment (pomegranited) [reply openassessment/data.py]: Perfect!

**Comment by swalladge** [comment]
@pomegranited Thanks :)

hehe oops I just realized that for the translations it had crawled everything in my virtualenv directory too. Don't have to worry about that now the translations are reverted. :sweat_smile:

**Comment by swalladge** [comment]
@natabene This is ready for review by Edx. :)

**Comment by natabene** [comment]
@ormsbee This is ready for your review.

**Comment by swalladge** [comment]
@ormsbee 

>  Do we know roughly how long this thing runs and how many queries it needs to fire off per workflow?

In testing (by adding `with self.assertNumQueries(1):` to `test_collect_ora2_summary`) it appears to send 8 database queries per submission. Compare to the `collect_ora2_data` which sends 2 queries per submission.

I'm not sure about speed; we haven't tried benchmarking this on a large instance.

This is naturally going to be heavier than `collect_ora2_data` since it aggregates more statistics. It is highly likely that it can be optimized though - queries aren't batched because the method uses api methods that each send a database query themselves.

**Comment by swalladge** [comment]
@ormsbee Do you know when we might get any updates on this? :)

**Comment by swalladge** [comment]
@natabene CC @ormsbee any updates on this? :slightly_smiling_face:

**Comment by natabene** [comment]
@swalladge No, no updates. Please note that I know prioritize all OpenCraft OSPRs with @pomegranited regularly, so you can get updates and help prioritize with her directly.

**Comment by marcotuts** [comment]
Would it be possible to get screenshots of where this CSV is beign shown / exposed?

**Comment by swalladge** [comment]
@marcotuts screenshots and example csv data added to the PR description. :slightly_smiling_face:

**Comment by swalladge** [comment]
Note: rebased on master and squashed

**Comment by nasthagiri** [comment]
@jnlapierre Here's another OSPR that may be related to the work your team is already engaged with. If not, you may need to discuss these changes with your PM.

**Comment by pomegranited** [comment]
Hi @natabene  -- we're doing a 2nd OpenCraft review on this PR during the next couple of weeks, so if that passes, I could technically merge this as a core committer.

But [Dave had some concerns](https://github.com/edx/edx-ora2/pull/1250#issuecomment-514935846), and [so did Marco](https://github.com/edx/edx-ora2/pull/1250#issuecomment-553184225), but we've addressed them and have had no reply. So I'm not sure whether it's ok to merge this.

What do you think @nedbat  and @feanil ?

**Comment by natabene** [comment]
@marcotuts Would you be ok with this merged?

**Comment by eLRuLL** [comment]
@swalladge looks good to me, sorry I didn't approve it appropiately before

- [x] I tested this with my local devstack
- [x] I read through the code and found it clean without problems
- [x] Includes documentation/comments

**Comment by pomegranited** [comment]
Thanks for your review @eLRuLL ! @swalladge  could you fix the conflict in setup.py and squash commits? Once we get @marcotuts 's OK, then I can merge. CC @natabene

**Comment by swalladge** [comment]
Conflicts fixed and commits squashed. :) @pomegranited

**Review by jnlapierre — APPROVED** [review]
LGTM , please resolve conflicts and prepare for merge

**Comment by swalladge** [comment]
@jnlapierre thanks!  I rebased and resolved conflicts. :)

**Comment by swalladge** [comment]
@jnlapierre @nasthagiri regarding the earlier conversation about how this fits with recent work. We had an internal conversation and determined that this report output does not overlap with other reports and api endpoints developed.  There are a two features which were developed which could be used to enhance this report though:

- a flag `ENABLE_ORA_USERNAMES_ON_DATA_EXPORT` was added in https://github.com/edx/edx-platform/pull/24198, which could be used to toggle showing real usernames on the report, instead of the anonymised usernames
- use the method added in https://github.com/edx/edx-ora2/pull/1438 to show problem block names in addition to the block ids

Please let us know if you would like us to use these features in this report as well, or if it's ok as-is. Thanks.

CC @clemente @giovannicimolin

**Comment by pomegranited** [comment]
@nasthagiri @jnlapierre I've flagged this for architecture review to address @swalladge 's comments above. CC @natabene 

Please let us know if it's ok to merge this?

**Comment by natabene** [comment]
@pomegranited Code Freeze is over, I think it is ok to merge, unless @nasthagiri or @jnlapierre have any objections.

**Comment by pomegranited** [comment]
Hi @jnlapierre , do you have any objections to this change?

**Comment by natabene** [comment]
@swalladge Could you please resolve the conflict?

**Comment by swalladge** [comment]
@natabene conflict resolved. :)

**Comment by natabene** [comment]
@swalladge Thanks, could you look into the failing check?

**Comment by natabene** [comment]
@swalladge Thank you!

File: openassessment/assessment/test/test_peer.py
Comment (swalladge) [inline openassessment/assessment/test/test_peer.py:1211]: I don't know what has changed here :thinking:

Comment (swalladge) [reply openassessment/assessment/test/test_peer.py]: Well CI didn't like the change, so maybe a quirk on my stack. :thinking:

**Comment by swalladge** [comment]
@natabene tests are all green again. :)

**Comment by natabene** [comment]
@jnlapierre Just checking if this is good to merge now.

**Comment by pomegranited** [comment]
Nusge @jnlapierre  ^ :)

**Comment by mattcarter** [comment]
resolved conflict in `setup.py`

**Comment by edx-webhook** [comment]
@swalladge 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-ora2&entry.867055334=https://github.com/edx/edx-ora2/pull/1250&entry.1484655318=https://github.com/swalladge&entry.752974735=2019-07-08+01:44&entry.1917517419=2021-01-28+03:17&entry.2133058324=Yes) so we can improve your experience in the future.

**Comment by swalladge** [comment]
Thanks @jnlapierre @mattcarter !

I updated [edx/edx-platform#20955](https://github.com/edx/edx-platform/pull/20955) to fix conflicts and use the new ora2 revision (which will work once 3.0.1 has been pushed to pypi).

**Pull request merged**
